### ISSUE TITLE
DNS lookups from Nginx to CoreDNS in eBPF mode fail when the CoreDNS Pods move

### ISSUE THREAD

**Issue by diranged** [projectcalico/calico#4509]
Slack Thread: https://calicousers.slack.com/archives/CPEPF833L/p1614095727031700

We discovered that when Nginx is doing DNS lookups, it uses the same source-port for all of its DNS queries. When operating in eBPF mode, Calico is doing the "service-to-ip" mapping the first time a connection is created, and then keeps that data for the duration of that particular connection. Because Nginx is using the same source port each time, Calico tracks this as the same connection. If your backing pods (coredns in this case) move, their IPs are never updated as part of this connection. This causes Nginx to fail to execute any future DNS lookups, breaking Nginx in many different ways.

## Expected Behavior

If the backing pod for an already-created connection session has been gone away, I think that any caching that calico does should get wiped out so that when a new session is created (even if its from the same source port), a new mapping is created.

## Steps to Reproduce (for bugs)

See the slack thread above for more details...

1. Create a cluster running with eBPF mode.
2. Create a service "guestbook"
3. Create a service "guestbook-proxy" that is an nginx reverse proxy pointing to "guestbook.<namespace>".
    (Make sure to set Nginx up to do DNS lookups against the coredns service)
4. `kubectl rollout restart deployment/coredns -n kube-system`
5. `kubectl rollout restart deployment/guestbook`

At this point, the `guestbook-proxy` pod should be unable to lookup the backend address for the `guestbook` service. It will report DNS lookup timeouts.

## Context
We make heavy use of reverse-proxies in our environment.. not only for our own apps, but even for Loki (the "loki-gateway" pod is an Nginx reverse proxy). We've had several outages in our staging environment due to our Coredns pods moving IPs and DNS resolution breaking.

## Your Environment
Kubernetes: EKS 1.19
Calico: 3.18.2
OS: Bottlerocket v1.0.7 (pre-release)

### tcpdump of the DNS lookups that work

```IP (tos 0x0, ttl 255, id 34455, offset 0, flags [DF], proto UDP (17), length 63)
    100.64.99.171.53869 > 100.64.147.127.53: [bad udp cksum 0xbfe7 -> 0xa0a3!] 44873+ A? foo.okta.com. (35)
IP (tos 0x0, ttl 254, id 34455, offset 0, flags [DF], proto UDP (17), length 63)
    100.64.99.171.53869 > 100.64.147.127.53: [bad udp cksum 0xbfe7 -> 0xa0a3!] 44873+ A? foo.okta.com. (35)
IP (tos 0x0, ttl 255, id 34456, offset 0, flags [DF], proto UDP (17), length 63)
    100.64.99.171.53869 > 100.64.147.127.53: [bad udp cksum 0xbfe7 -> 0x5833!] 56505+ AAAA? foo.okta.com. (35)
IP (tos 0x0, ttl 254, id 34456, offset 0, flags [DF], proto UDP (17), length 63)
    100.64.99.171.53869 > 100.64.147.127.53: [bad udp cksum 0xbfe7 -> 0x5833!] 56505+ AAAA? foo.okta.com. (35)
IP (tos 0x0, ttl 254, id 58486, offset 0, flags [DF], proto UDP (17), length 329)
    100.64.147.127.53 > 100.64.99.171.53869: [udp sum ok] 56505 q: AAAA? foo.okta.com. 2/1/0 foo.okta.com. CNAME ok3-crtrs.tng.okta.com., ok3-crtrs.tng.okta.com. CNAME ok3-crtr-tls12-nlb-dfef298ffc8f82ca.elb.us-east-1.amazonaws.com. ns: elb.us-east-1.amazonaws.com. SOA ns-1826.awsdns-36.co.uk. awsdns-hostmaster.amazon.com. 1 7200 900 1209600 60 (301)
IP (tos 0x0, ttl 253, id 58486, offset 0, flags [DF], proto UDP (17), length 329)
    100.64.147.127.53 > 100.64.99.171.53869: [udp sum ok] 56505 q: AAAA? foo.okta.com. 2/1/0 foo.okta.com. CNAME ok3-crtrs.tng.okta.com., ok3-crtrs.tng.okta.com. CNAME ok3-crtr-tls12-nlb-dfef298ffc8f82ca.elb.us-east-1.amazonaws.com. ns: elb.us-east-1.amazonaws.com. SOA ns-1826.awsdns-36.co.uk. awsdns-hostmaster.amazon.com. 1 7200 900 1209600 60 (301)
IP (tos 0x0, ttl 254, id 58487, offset 0, flags [DF], proto UDP (17), length 452)
    100.64.147.127.53 > 100.64.99.171.53869: [udp sum ok] 44873 q: A? foo.okta.com. 5/0/0 foo.okta.com. CNAME ok3-crtrs.tng.okta.com., ok3-crtrs.tng.okta.com. CNAME ok3-crtr-tls12-nlb-dfef298ffc8f82ca.elb.us-east-1.amazonaws.com., ok3-crtr-tls12-nlb-dfef298ffc8f82ca.elb.us-east-1.amazonaws.com. A 18.209.113.152, ok3-crtr-tls12-nlb-dfef298ffc8f82ca.elb.us-east-1.amazonaws.com. A 18.209.113.153, ok3-crtr-tls12-nlb-dfef298ffc8f82ca.elb.us-east-1.amazonaws.com. A 18.209.113.160 (424)
IP (tos 0x0, ttl 253, id 58487, offset 0, flags [DF], proto UDP (17), length 452)
    100.64.147.127.53 > 100.64.99.171.53869: [udp sum ok] 44873 q: A? foo.okta.com. 5/0/0 foo.okta.com. CNAME ok3-crtrs.tng.okta.com., ok3-crtrs.tng.okta.com. CNAME ok3-crtr-tls12-nlb-dfef298ffc8f82ca.elb.us-east-1.amazonaws.com., ok3-crtr-tls12-nlb-dfef298ffc8f82ca.elb.us-east-1.amazonaws.com. A 18.209.113.152, ok3-crtr-tls12-nlb-dfef298ffc8f82ca.elb.us-east-1.amazonaws.com. A 18.209.113.153, ok3-crtr-tls12-nlb-dfef298ffc8f82ca.elb.us-east-1.amazonaws.com. A 18.209.113.160 (424)
```

### tcpdump from the same host, after coredns has moved
```
IP (tos 0x0, ttl 255, id 12183, offset 0, flags [DF], proto UDP (17), length 63)
    100.64.99.171.53869 > 100.64.147.127.53: [bad udp cksum 0xbfe7 -> 0x6a0f!] 58845+ A? foo.okta.com. (35)
IP (tos 0x0, ttl 254, id 12183, offset 0, flags [DF], proto UDP (17), length 63)
    100.64.99.171.53869 > 100.64.147.127.53: [bad udp cksum 0xbfe7 -> 0x6a0f!] 58845+ A? foo.okta.com. (35)
IP (tos 0x0, ttl 255, id 12184, offset 0, flags [DF], proto UDP (17), length 63)
    100.64.99.171.53869 > 100.64.147.127.53: [bad udp cksum 0xbfe7 -> 0xb160!] 33676+ AAAA? foo.okta.com. (35)
IP (tos 0x0, ttl 254, id 12184, offset 0, flags [DF], proto UDP (17), length 63)
    100.64.99.171.53869 > 100.64.147.127.53: [bad udp cksum 0xbfe7 -> 0xb160!] 33676+ AAAA? foo.okta.com. (35)
IP (tos 0x0, ttl 255, id 13049, offset 0, flags [DF], proto UDP (17), length 63)
    100.64.99.171.53869 > 100.64.147.127.53: [bad udp cksum 0xbfe7 -> 0x6a0f!] 58845+ A? foo.okta.com. (35)
IP (tos 0x0, ttl 254, id 13049, offset 0, flags [DF], proto UDP (17), length 63)
    100.64.99.171.53869 > 100.64.147.127.53: [bad udp cksum 0xbfe7 -> 0x6a0f!] 58845+ A? foo.okta.com. (35)
IP (tos 0x0, ttl 255, id 13050, offset 0, flags [DF], proto UDP (17), length 63)
    100.64.99.171.53869 > 100.64.147.127.53: [bad udp cksum 0xbfe7 -> 0xb160!] 33676+ AAAA? foo.okta.com. (35)
IP (tos 0x0, ttl 254, id 13050, offset 0, flags [DF], proto UDP (17), length 63)
    100.64.99.171.53869 > 100.64.147.127.53: [bad udp cksum 0xbfe7 -> 0xb160!] 33676+ AAAA? foo.okta.com. (35)
IP (tos 0x0, ttl 255, id 13824, offset 0, flags [DF], proto UDP (17), length 63)
    100.64.99.171.53869 > 100.64.147.127.53: [bad udp cksum 0xbfe7 -> 0x6a0f!] 58845+ A? foo.okta.com. (35)
IP (tos 0x0, ttl 254, id 13824, offset 0, flags [DF], proto UDP (17), length 63)
    100.64.99.171.53869 > 100.64.147.127.53: [bad udp cksum 0xbfe7 -> 0x6a0f!] 58845+ A? foo.okta.com. (35)
IP (tos 0x0, ttl 255, id 13825, offset 0, flags [DF], proto UDP (17), length 63)
    100.64.99.171.53869 > 100.64.147.127.53: [bad udp cksum 0xbfe7 -> 0xb160!] 33676+ AAAA? foo.okta.com. (35)
IP (tos 0x0, ttl 254, id 13825, offset 0, flags [DF], proto UDP (17), length 63)
    100.64.99.171.53869 > 100.64.147.127.53: [bad udp cksum 0xbfe7 -> 0xb160!] 33676+ AAAA? foo.okta.com. (35)
IP (tos 0x0, ttl 255, id 59601, offset 0, flags [DF], proto UDP (17), length 81)
```

**Comment by tomastigera** [comment]
The root cause seems to be the connect time load balancer. Setting `FELIX_BPFCONNECTTIMELOADBALANCINGENABLED` to `false` resolves the issue - but causes issues for host-networked pods as described in the conf option description.

We make the backend choice once at `conect()` time. If the backend disappears, even if we delete conntrack entries for the now missing backend, the socket does not pick up the change and keeps trying to send packets to the now-missing backend. Unless the application closes the socket due to errors, there is no way for the dataplane to resolve this. Closing and recreating the socket would pick a new existing backend.

Disabling connect time LB means that we make a new LB decision each time we have a conntrack miss, therefore the changing the config option resolves the issue - however, this is a slower mode as we ned to NAT each packet and it causes other issues :-/

**Comment by diranged** [comment]
@tomastigera, @lwr20 Any chance this is going to be looked at soonish?

**Comment by diranged** [comment]
Any updates on this issue?

**Comment by diranged** [comment]
@tomastigera So I’ve been wondering - how do others deal with this? How are people running eBPF mode and dealing with the fact that your CoreDNS pods might move around. That leads me to this question: Should we be changing up our DNS configuration by implementing something like https://kubernetes.io/docs/tasks/administer-cluster/nodelocaldns/ to route DNS requests to CoreDNS over TCP which would then bypass this issue? Or is there another architecture we should be thinking about?

**Comment by salanki** [comment]
@diranged: I am dealing with it by having livenessProbes that trigger on the NGINX dns resolution failure and kills the pod. It is a super hack of course.

**Comment by tomastigera** [comment]
I am actively working on resolving this issue properly. We want to use the benefit of connect time load balancing as much as possible so we want to use the slower path only when necessary. There is a whole bunch of scenarios. Stay tuned :pray: it may be a :christmas_tree: :package:

**Comment by tomastigera** [comment]
Just a short update on this issue. We have a [PoC](https://github.com/projectcalico/calico/pull/5357) but we are not quite happy with that yet :man_shrugging:

**Comment by diranged** [comment]
Just checking back in... any further progress?

**Comment by tomastigera** [comment]
This is still being worked on

**Comment by diranged** [comment]
Thanks... I really hope we (hah, you guys) can figure it out! :)

**Comment by tomastigera** [comment]
@diranged @salanki if you set `FeatureGates` on current master, for instance env `FELIX_FeatureGates`, to equal `BPFConnectTimeLoadBalancingWorkaround=udp` then it should disable CTLB only for UDP and allow host networked processes to use udp services without CTLB (previously not possible without CTLB)

This in turn will eventually lead to cleaning up conntrack entries without a backend and that would make your pods to try another backend. Therefore if DNS pods migrated, pods will try new DNS backend and so it will resolve the DNS issue.

It is still in a testing / non-GA phase, hence the gate. Please, could you try it out and let us know whether things work for you. It took a long time to get it to this stage, so I hope it will help to resolve your issues! :crossed_fingers:

**Comment by tomastigera** [comment]
> At least in my scenario, nothing is not networked. Both the client and CoreDNS is on the calico pod network. Does that matter for the workaround?

Previously there was either CTLB for everything (higher perf, sometimes broken DNS) or no CTLB (lower perf, DNS worked, host networked did not work at at all).

This adds the option if you want to keep CTLB, which avoids NATing every packet on service connections, for TCP. This setting will only disable CTLB for udp and makes sure that it also works for host networked, which you don't care about, but others may :)

If you care about CTLB for UDP, e.g. you use it beyond DNS, say for data transfers, you may feel some perf impact. In that case you would need to stick with your live-probes. We may also refine this and let you select UDP service or service port that you care about, say enable this workaround just for DNS. But that is not available yet.

**Comment by salanki** [comment]
Writing to note that this issue still exists.

**Comment by tomastigera** [comment]
The issue is considered fixed since 3.26 and the fix is part of the default setting in 3.27

**Comment by zoftdev** [comment]
> The issue is considered fixed since 3.26 and the fix is part of the default setting in 3.27
@tomastigera 
I see this commit has been merge to 3.27 and v3.28.0-0.dev
https://github.com/projectcalico/calico/pull/8189
Did it already fix 3.26?

**Comment by zoftdev** [comment]
> The issue is considered fixed since 3.26 and the fix is part of the default setting in 3.27
@tomastigera  I check commit  https://github.com/projectcalico/calico/pull/8189  , it merge to 3.27 and v3.28.0-0.dev

Dose it work on 3.26?

**Comment by tomastigera** [comment]
@zoftdev shoudl work in 3.26 with these settings https://github.com/projectcalico/calico/issues/4509#issuecomment-1306318160

**Comment by zoftdev** [comment]
> The issue is considered fixed since 3.26 and the fix is part of the default setting in 3.27

@tomastigera 
I tested version 3.27 to see if the commit could resolve the Istio compatibility issue. However, I encountered a connection timeout when the client pod was using host networking (e.g., a calico-controller node) and the destination was a service with an external IP (e.g., an API server installed outside the pod network).

This is endpoint which have problem.
> 
> apiVersion: v1
> kind: Endpoints
> metadata:
>   name: kubernetes
>   namespace: default
> subsets:
> - addresses:
>   - ip: 10.10.66.11
>   - ip: 10.10.66.12
>   - ip: 10.10.66.13
>   ports:
>   - name: https
>     port: 6443
>     protocol: TCP
>  

In another case (  host-network ->svc , pod->svc of pod ) is ok.

**Comment by tomastigera** [comment]
@zoftdev lets take this to a new issue
